### PR TITLE
Add raw java example for spi-based-sharding-algorithm-example

### LINKED ISSUE #15717: Add raw java example for `spi-based-sharding-algorithm-example`
Hi community,

### Background

[spi-based-sharding-algorithm-example](https://github.com/apache/shardingsphere/tree/master/examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example) is an example of a custom algorithm that demonstrates how to integrate a user-defined sharding algorithm and configure it using  [SPI (Service Provider Interface)](https://docs.oracle.com/javase/tutorial/sound/SPI-intro.html).

### Aim

Add raw java example for `spi-based-sharding-algorithm-example` (in sub-module `spi-based-sharding-raw-jdbc-example`)

### Basic Qualifications

- Java
- Maven

### Example FYI

- [sharding-raw-jdbc-example](https://github.com/apache/shardingsphere/tree/master/examples/shardingsphere-jdbc-example/single-feature-example/sharding-example/sharding-raw-jdbc-example)

### PR DESCRIPTION
**Pull request by lushaorong**
Fixes #15717.

Changes proposed in this pull request:

- Add raw java example for spi-based-sharding-algorithm-example

### CODE REVIEW

File: examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/config/SPIBasedShardingRawJavaConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        result.getBindingTableGroups().add("t_order, t_order_item"); 
+        result.getBroadcastTables().add("t_address");
+        result.setDefaultDatabaseShardingStrategy(new StandardShardingStrategyConfiguration("user_id", "spi_based"));
+        result.getShardingAlgorithms().put("spi_based", new AlgorithmConfiguration("SPI_BASED", new Properties()));
+        result.getKeyGenerators().put("snowflake", new AlgorithmConfiguration("SNOWFLAKE", new Properties()));
+        return result;
+    }
+    
+    private ShardingTableRuleConfiguration getOrderTableRuleConfiguration() {
+        ShardingTableRuleConfiguration result = new ShardingTableRuleConfiguration("t_order", "demo_ds_${0..1}.t_order_${[0, 1]}");
```
Comment (RaigorJiang) [inline examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/config/SPIBasedShardingRawJavaConfiguration.java:63]: Hi @lushaorong,
Nice to see you again, here's a little problem:
The sharding type you specified is `SHARDING_DATABASES`, but the actualDataNodes here contains the logic of sharding tables, which is obviously incorrect.

It is recommended that you test it before submitting a PR. If you have any questions, welcome to communicate.

Comment (lushaorong) [reply examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/config/SPIBasedShardingRawJavaConfiguration.java:63]: SHARDING_DATABASES_AND_TABLES is more suitable? 
But I refer to [here](https://github.com/apache/shardingsphere/blob/master/examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/SPIBasedShardingRawYamlConfigurationExample.java), it contains the logic of sharding tables and use SHARDING_DATABASES.Is it right？

Comment (RaigorJiang) [reply examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/config/SPIBasedShardingRawJavaConfiguration.java:63]: Either way, the configuration should match the name.
But you used `SHARDING_DATABASES`, and dataNodes `"demo_ds_${0..1}.t_order_${[0, 1]}"` is for `SHARDING_DATABASES_AND_TABLES`.

This is not a match.

Comment (lushaorong) [reply examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/config/SPIBasedShardingRawJavaConfiguration.java:63]: You're right.
I'll change it to SHARDING_DATABASES_AND_TABLES

Comment (lushaorong) [reply examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/config/SPIBasedShardingRawJavaConfiguration.java:63]: It is still SHARDING_DATABASES and I change the rule `demo_ds_${0..1}.t_order_item_${[0, 1]}` to `demo_ds_${0..1}.t_order_item`

**Review by RaigorJiang — APPROVED** [review]

**Pull request merged**